### ISSUE TITLE
Issues implementing consult-dash

### ISSUE THREAD

**Issue by orki** [minad/consult#534]
Here's my attempt at implementing a consult interface for dash docs; this is loosely modeled on consult-spotify, but have run into a couple of issues. I have converted the [synchronous function](https://github.com/dash-docs-el/dash-docs/blob/master/dash-docs.el#L571) `dash-docs-search` into an asynchronous function:
```
(defvar consult-dash--results nil
  "Stores the previously retrieved docset results")
(defun consult-dash--collection (callback search-string &rest _)
  (setq consult-dash--results (dash-docs-search search-string))
  (funcall callback consult-dash--results))
;; Asynchronous interface to dash-docs for use with consult
(defun consult-dash--async-query (callback search-string)
  (run-at-time 0 nil #'consult-dash--collection callback search-string))
```
1. Running `consult-dash--async-query` seems to work only outside `consult--read`; inside `consult--read`, `(dash-docs-search search-string)` always returns `nil` . I'm sure I'm doing something boneheaded with async processes since `dash-docs-search` internally uses `call-process`, but I'm not sure how to debug. Eliminating the temporary `consult-dash--results` above by `(funcall callback (dash-docs-search search-string))` does not help.
2. If the previous issue can be resolved, are string properties the right way to attach metadata to completion candidates? Completion candidates are of the form `(list "some text" "property1" "property2")`; once `consult--read` is done, it would be useful to get the whole list, and not just `"some text"`.

The entirety of the code (including debugging `message` statements) is reproduced below:
```
(require 'dash-docs)
(require 'cl-lib)
(require 'subr-x)

(defvar-local consult-dash-docsets nil
  "Docsets to use for this buffer")

(defvar consult-dash--results nil
  "Stores the previously retrieved docset results")
(advice-add #'dash-docs-buffer-local-docsets :around
              (lambda (old-fun &rest args)
                (let ((old (apply old-fun args)))
                  (cl-remove-duplicates (append old consult-dash-docsets)))))

(defun consult-dash--collection (callback search-string &rest _)
  ;; (message (prin1-to-string (dash-docs-buffer-local-docsets)))
  (message "Searching for \"%s\" " search-string)
  (message (concat "Results were:" (prin1-to-string (dash-docs-search search-string))))
  (setq consult-dash--results (dash-docs-search search-string))
  ;; (message (concat "Results were:" (prin1-to-string consult-dash--results)))
  (funcall callback consult-dash--results))
  ;; (funcall callback (mapcar (lambda (x) (list (concat search-string x) (downcase x))) '("AA" "AB" "CC" "D" "BE"))))
(defun consult-dash--format-item (item)
  "Format ITEM as completion candidate"
  (car item))

;; Asynchronous interface to dash-docs for use with consult
(defun consult-dash--async-query (callback search-string)
  (run-at-time 0 nil #'consult-dash--collection callback search-string))

(defun consult-dash--browse-matching-result (match)
  "Given a MATCH, find matching result and browse it's url."
  (when-let ((result
              (cdr (cl-find-if (lambda (e)
                                 (string= match (car e)))
                               consult-dash--results))))
    (dash-docs-browse-url result)))

;; Debugging code for now
;; (progn
;;   (setq consult-dash-docsets '("C" "C++" "Boost" "Qt"))
;;   (defvar consult-dash--choices nil "")
;;   (consult-dash--async-query (lambda (val) (setq consult-dash--choices val)) "t32_t")
;;   (sit-for 0)
;;   consult-dash--choices)
;; End of debugging code

(defvar consult-dash-history nil
  "Previous queries for dash docs via consult")
(defun consult-dash-with-args ()       ; currently no args; will add once done
  "Consult dash with some args"
  (consult--read (consult-dash--search-generator)
                 :prompt "Dash search: " ; to do: display list of docsets
                 :lookup #'consult--lookup-member
                 :category 'dash-search-item
                 :history '(:input consult-dash-history)
                 :initial (consult--async-split-thingatpt 'symbol)
                 :require-match t))
(defun consult-dash--search-generator () ; currently no args; will add once done
  (thread-first (consult--async-sink)
    (consult--async-refresh-immediate)
    (consult--async-map #'consult-dash--format-item)
    (consult-dash--async-search)
    ;; (consult--async-throttle)
    (consult--async-split)))

(defun consult-dash--async-search (next)
  "Async search with NEXT"
  (let ((current ""))
    (lambda (action)
      (pcase action
        ((pred stringp)
         ;; (message "Got a string: %s, previous %s" action current)
         (when (and (> (length action) 0)
                    (or (< (length current) dash-docs-min-length)
                        (not (string-prefix-p current action))))
           (setq current action)
           (consult-dash--async-query
            (lambda (candidates)
              ;; (message "Got candidates: %s" (prin1-to-string candidates))
              (funcall next 'flush)
              (funcall next candidates))
            action)))
        (_ (funcall next action))))))

(defun consult-dash-test ()
  (interactive)
  (message (concat "Consult gave: " (prin1-to-string (consult-dash-with-args)))))

(provide 'consult-dash)
```

**Comment by minad** [comment]
You should use a different approach to fetch the asynchronous candidates, where the candidates produced by the background process are generated via a callback. There are multiple examples you can look at: consult-notmuch, consult-spotify, consult-recoll, affe. Currently you still generate the candidates synchronously in an idle timer, which will not work.

Maybe https://github.com/astoff/devdocs.el is also interesting for you?

**Comment by orki** [comment]
> You should use a different approach to fetch the asynchronous candidates, where the candidates produced by the background process are generated via a callback. There are multiple examples you can look at: consult-notmuch, consult-spotify, consult-recoll, affe. Currently you still generate the candidates synchronously in an idle timer, which will not work.

In that case, reimplementing a significant fraction of [dash docs](https://github.com/dash-docs-el/dash-docs) inside consult-dash cannot be avoided. I'll update/reopen this issue once I make more progress; this issue can be closed until then.

> Maybe https://github.com/astoff/devdocs.el is also interesting for you?

It is interesting, but I also have quite a few datasets already generated in the dash sqlite format, which will require some additional work to regenerate.

**Comment by minad** [comment]
I suggest you ask the authors of dash-docs to provide an async api. Maybe they are interested too in a Consult support package?

**Comment by orki** [comment]
Managed to implement it with the code at the end of this comment. The biggest question is to find a way to separate matches from each docset into its own section; current code:
```
(defun consult-dash--format (lines)
  (let ((candidates)
        (current-candidate))
    (save-match-data
      (dolist (str lines)
        (setq current-candidate (split-string str "|" t))
        (if (= 1 (length current-candidate))
            ;; FIXME: If we do not find the right prefix, we should raise an error
            (when (string-prefix-p consult-dash--docset-prefix str)
              (setq consult-dash--current-docset (substring str (length consult-dash--docset-prefix))))
          (let ((name (cadr current-candidate))
                (type (car current-candidate)))
            (add-face-text-property 0 (length name) 'consult-key nil name)
            (push (list
                   (format "%s: %s (%s)" consult-dash--current-docset name type)
                   consult-dash--current-docset
                   current-candidate)
                  candidates)))))
    (nreverse candidates)))
```
Currently each candidate is listed as `Docset Name: topic (type)`. How do I collate the results for each docset into its own "section" with the name "Docset Name" in the example above? `consult-grep` seems to create these sections per file. Entirety of code follows:
```
;;; consult-dash.el --- Consult front-end for dash-docs  -*- lexical-binding: t; -*-

;; Copyright (C) 2022  Ravi R Kiran

;; Author: Ravi R Kiran <aine.marina@gmail.com>
;; Keywords: consult, dash

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; consult-dash is the only interface function

;; To do

;; - Segment matches per docset
;; - Avoid concatenating commands through the shell

;;; Code:

(require 'dash-docs)
(require 'subr-x)

(defvar-local consult-dash-docsets nil
  "Docsets to use for this buffer")
(defvar consult-dash-sqlite-args "sqlite3 -list"
  "Sqlite command-line arguments")

(advice-add #'dash-docs-buffer-local-docsets :around
              (lambda (old-fun &rest args)
                (let ((old (apply old-fun args)))
                  (delq nil (delete-dups (append old consult-dash-docsets))))))

(defvar consult-dash--history nil
  "Previous queries for dash docs via consult")
(defvar consult-dash--docset-prefix "DOCSET:"
  "Prefix to identify docset boundaries")

(defun consult-dash--builder-one-docset (docset pattern)
  (let* ((query (dash-docs-sql-query (caddr docset)
                                     (dash-docs-sub-docset-name-in-pattern pattern
                                                                           (car docset))))
         (cmd (concat "echo " consult-dash--docset-prefix
                      (shell-quote-argument (car docset))
                      "; "
                      consult-dash-sqlite-args
                      " "
                      (shell-quote-argument (cadr docset))
                      " "
                      (shell-quote-argument query)
                      "; ")))
    cmd))

;; Is there a better way to run multiple commands when using consult--async-command?
(defun consult-dash--builder (input)
  (pcase-let ((`(,arg . ,opts) (consult--command-split input)))
    (unless (string-blank-p arg)
      (when-let* ((docsets (dash-docs-maybe-narrow-docsets arg))
                  (cmds (mapcar (lambda (ds) (consult-dash--builder-one-docset ds arg)) docsets)))
        (list :command (list "sh" "-c" (apply #'concat cmds))
              :highlight (cdr (consult--default-regexp-compiler arg 'basic)))))))

(defun consult-dash--with-buffer-context (func)
  "Ensure that FUNC is called with the correct buffer context"
  (let ((buf (current-buffer)))
    (lambda (&rest args)
      (with-current-buffer buf
        (apply func args)))))

(defvar consult-dash--current-docset nil
  "Stored current docset output for chunked calls to consult-dash--format")

(defun consult-dash--format (lines)
  (let ((candidates)
        (current-candidate))
    (save-match-data
      (dolist (str lines)
        (setq current-candidate (split-string str "|" t))
        (if (= 1 (length current-candidate))
            ;; FIXME: If we do not find the right prefix, we should raise an error
            (when (string-prefix-p consult-dash--docset-prefix str)
              (setq consult-dash--current-docset (substring str (length consult-dash--docset-prefix))))
          (let ((name (cadr current-candidate))
                (type (car current-candidate)))
            (add-face-text-property 0 (length name) 'consult-key nil name)
            (push (list
                   (format "%s: %s (%s)" consult-dash--current-docset name type)
                   consult-dash--current-docset
                   current-candidate)
                  candidates)))))
    (nreverse candidates)))

(defun consult-dash (&optional initial)
  (interactive)
  (dash-docs-create-common-connections)
  (dash-docs-create-buffer-connections)
  (setq consult-dash--current-docset nil)
  (when-let* ((builder (consult-dash--with-buffer-context #'consult-dash--builder))
              (search-result (consult--read
                              (consult--async-command builder
                                (consult--async-transform consult-dash--format)
                                (consult--async-highlight builder))
                              :prompt "Dash: "
                              :require-match t
                              :lookup #'consult--lookup-cdr
                              :initial (consult--async-split-initial initial)
                              :add-history (consult--async-split-thingatpt 'symbol)
                              :history '(:input consult-dash--history))))
    (dash-docs-browse-url search-result)))

(provide 'consult-dash)
;;; consult-dash.el ends here
```

**Comment by oantolin** [comment]
You should use the `:group` keyword in the `consult--read` call to add a `group-function`. The group function should be a function of two arguments, `candidate` and `transform`. When `transform` is nil, the function should return the name of the group `candidate` should be placed in. When `tranform` is non-nil the function should instead return the way you want the candidate displayed in its group, in this case, it should the `Docset Name:` part from `candidate`.

**Comment by orki** [comment]
Thanks, that helps [the implementation quite a bit](http://cgit.lexarcana.com/cgit.cgi/dotemacs/commit/?id=f79eb3de1cb18941f22e512853d983cea3c00c8f).

The next step is to figure out a solution for users without an `sh`; perhaps `consult-multi` can be made to work with asynchronous sources. Once this is done, I'll probably split this into a standalone package and release it into the wild.

**Comment by minad** [comment]
> perhaps consult-multi can be made to work with asynchronous sources.

Asynchronous multi sources are out of scope. This feature has not been widely requested. I consider it an unnecessary complication which leads to many issues (performance, treatment of asynchronous input etc). I also doubt that it would be useful in many scenarios. See #196 for a previous discussion.

Do you have more questions here? I am looking forward to seeing it published on MELPA!

**Comment by orki** [comment]
[Published it](https://codeberg.org/ravi/consult-dash); will add to MELPA after at least one or two others test it.

**Comment by minad** [comment]
You put up a few questions in https://runes.lexarcana.com/2022-04-16-consult-interface-for-dash-docs.html.

> Libraries must use an asynchronous generator of completion results to emulate dynamic completion tables with consult. This approach is much nicer in terms of lag, but complicates library design significantly. 

First a comment about this: You don't have to use the consult async facilities if you don't want the asynchronicity. Why not use use plain synchronous `completing-read` then if that is what you want?

~~~ elisp
(let ((completion-styles '(basic)))
  (completing-read
   "Prompt: "
   (completion-table-dynamic (lambda (input)
                               (list (format "%s1" input)
                                     (format "%s2" input)
                                     (format "%s3" input))))))
~~~

I disagree that library design is complicated significantly. This is the equivalent using an asynchronous completion function. The main difference is that we have to capture the candidates in a closure. The asynchronous completion tables is not what complicates the design here. The actual complexity is inherent complexity of the process management, the advanced filtering, the debouncing and so on. There is a lot going on.

~~~ elisp
(consult--read
 (let (candidates)
   (lambda (input)
     (pcase input
       ((pred stringp)
        (setq candidates (list (format "%s1" input)
                         (format "%s2" input)
                         (format "%s3" input))))
       ('nil candidates)))))
~~~

Take a look at the protocol used by asynchronous sources for the details:

https://github.com/minad/consult/blob/a924b3b0d943d12c71d23964da7733251f4705b2/consult.el#L1676-L1715

> Please let me know if you have a way to avoid this global variable

Instead of a global variable you should capture the state in the closure. For example `consult--async-sink` captures the incoming candidates in its closure.

> Any ideas on ways to execute these command sequences with pure emacs facilities would also be greatly appreciated.

The problem is that you use the 'consult--async-command` functionality which builds upon a single `consult--async-process` source. The function `consult--async-process` manages a single process since this is all that was needed until now for the asynchronous Consult commands. If you want to manage multiple processes, start them in parallel, you have to write your own process management facilities. You can start multiple processes with `make-process`. Take a look at how the async command pipeline is put together:

https://github.com/minad/consult/blob/a924b3b0d943d12c71d23964da7733251f4705b2/consult.el#L1979-L1993

There you would need to replace `consult--async-process` with your own multi process manager and so on. Unfortunately it won't be trivial. In principle it should be possible to use multiple `consult--async-process` closures and merge the results together. The pipeline design lets you do all this in principle but there may be some practical limitations due to some assumptions of the pipeline components. For example `consult--async-process` also manages the status indicator in the minibuffer prompt, which one would have to factor out.

> consult--read itself returns the original non-formatted candidate. Any ideas on better techniques would also be greatly appreciated.

This is a fundamental limitation of `completing-read` which makes sense since `completing-read` is about completing a text string not necessarily about selecting a candidate. The command `exit-minibuffer` exits the minibuffer session and the minibuffer content is then used as return value. But the content of the minibuffer could be just some text entered by the user. For `consult-grep` I use `:lookup #'consult--lookup-member` to lookup a candidate string including its text properties. The only requirement is then that the candidate string is unique.

**Comment by orki** [comment]
You are right on all counts, of course. I edited the blog post to ensure that your comments are incorporated.

It came off abrasive: I apologize. My intent was to augment jao's posts on writing consult interfaces for other tools, by providing war stories to help others. In particular, I want to use `consult--read`, which is the reason for writing `consult-dash` in the first place. `consult` itself is an amazing piece of work on which I am excited to build tools. Thank you for helping with the implementation and for your patience.

**Comment by minad** [comment]
@orki Thanks. I am perfectly okay with a critical look at the API but I believe most of the limitations are inherited from the underlying `completing-read` API and even there the limitations make some sense. Of course if starting anew one would probably make a few different choices regarding the design of `completing-read` but that's not the goal here. Consult and related packages just try to work with the existing facilities to give the best possible integration, possibly making sacrifices elsewhere. If you have concrete proposals on how the `consult--read` API could be improved, I am all ears. 

Regarding dynamic completion tables I considered moving closer to the `completing-read` API, adding supporting for dynamic completion tables like `completion-table-dynamic` (see issue https://github.com/minad/consult/issues/319). I didn't do this yet since the design wasn't clear and there were also no clear advantages given that the Consult commands are either static or asynchronous. Furthermore `completion-table-dynamic` has significant issues, since it is not compatible with advanced completion styles. The only completion style which work are `basic`, `emacs21` and `emacs22`. Try this:

~~~ elisp
(let ((completion-styles '(flex))) ;; substring, partial-completion, initials, orderless all don't work.
  (completing-read
   "Prompt: "
   (completion-table-dynamic (lambda (input)
                               (list (format "%s1" input)
                                     (format "%s2" input)
                                     (format "%s3" input))))))
~~~

The combination with `basic` works too, but you don't profit from the advanced completion style then.

~~~ elisp
(let ((completion-styles '(orderless basic))) ;; or (basic orderless)
  (completing-read
   "Prompt: "
   (completion-table-dynamic (lambda (input)
                               (list (format "%s1" input)
                                     (format "%s2" input)
                                     (format "%s3" input))))))
~~~

Then there is another issue - Selectrum doesn't support `completion-table-dynamic` at all, not even with `basic`, but this is more a bug in Selectrum than an issue of the design of `completion-table-dynamic`. It has some historical significance, since when Consult started I was a Selectrum user and I wanted to have a maintainable set of utilities instead of just taking them from the [Wiki](https://github.com/raxod502/selectrum/wiki/Useful-Commands). These days one can use Vertico instead of Selectrum which fixes all the underlying architectural issues.

**Comment by orki** [comment]
I didn't realize that vertico supported dynamic completion tables. Coincidentally, I [switched from selectrum to vertico](http://cgit.lexarcana.com/cgit.cgi/dotemacs/commit/?id=ece9d8ece8e98782460da79ab0f2cf428e8f815a) only 40 hours ago, and not for dynamic completion tables.

**Comment by minad** [comment]
Isn't the Vertico README sufficiently clear about the reasons of Vertico's existence? Note that I contributed quite a bit to Selectrum before giving up, see https://github.com/raxod502/selectrum/graphs/contributors.
